bundles/org.openhab.binding.smarther/src/main/java/org/openhab/binding/smarther/internal/api/exception/SmartherAuthorizationException.java

```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.smarther.internal.api.exception;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * BTicino/Legrand API gateway authorization problems exception class.
+ *
+ * @author Fabio Possieri - Initial contribution
+ */
+@NonNullByDefault
+public class SmartherAuthorizationException extends RuntimeException {
```

Comment:
Is there a reason for extending from an unchecked exception? Would `extends Exception` be sufficient?

Reply:
I took the base skeleton from Spotify binding, where these were all extension of RuntimeException. Extending from Exception is not sufficient in this case: also trying to refactor them, I come into lot of troubles when dealing with CompletableFuture and ExpiringCache declared in framework methods (like initialize, etc.) where signature cannot be changed.

Reply:
I can't follow you. Can you point to a location that makes trouble?

Reply:
Just as an example: in the `initialize` function, I initialize the cache `locationCache` with action `getLocationList` that can throw `SmartherAuthenticationException`.

> locationCache = new ExpiringCache<>(Duration.ofMinutes(config.getStatusRefreshPeriod()), this::getLocationList);

Since this is an ExpiringCache, it is supposed to call `getLocationList` only when getting value from it, so in other functions where the exception are catched and handled (and I want to have exception thrown by `getLocationList` there, to be able to manage the related cases).

However, if I extend `SmartherAuthenticationException` from Exception I'm obliged to catch the exception even in `initialize` method, where it shouldn't be needed...

With "Spotify binding" I just meant I've started from there to build up my binding (as it handles OAuth2 authentication, also used by Legrand API) and, since it was an already reviewed and authorized binding, I thought it was a good starting point. That's it :)

Reply:
Understood, you used an unchecked exception, because you want to throw exceptions, but the external interface you used doesn't declare any `throws`. Well, the downside of this approach is that the compiler does not force you to catch the exception at the end.

I took another look at your code and found a location straightaway where you miss to catch an unchecked exception: `SmartherAuthorizationServlet.doGet()` invokes `formatApplications()` which eventually accesses your cache. In this case you may be lucky if an exception is thrown, the servlet thread would catch it for you (but what happens with the request?). In a plain Java application the current thread would be terminated.

I would catch the exception in `getLocationList`, log it and return an empty List or an empty Optional.

I am wondering what @cpmeister would say.